api/httpclient.go

```diff
@@ -89,44 +84,27 @@ func (doer httpRequestDoer) Do(req *http.Request) (*http.Response, error) {
 // AuthHTTPRequest adds Juju auth info (username, password, nonce, macaroons)
 // to the given HTTP request, suitable for sending to a Juju API server.
 func AuthHTTPRequest(req *http.Request, info *Info) error {
-	var tag string
+	var tag names.Tag
 	if info.Tag != nil {
-		tag = info.Tag.String()
+		tag = info.Tag
 	}
-	return authHTTPRequest(req, tag, info.Password, info.Nonce, info.Macaroons)
```

Comment:
This whole block can be deleted.
Just pass `info.Tag` to `NewUserpassLoginProvider` directly.

Reply:
Good catch